### ISSUE TITLE
CRITICAL: Update "joined", "left", "added", and "removed" behavior based on room type

### ISSUE THREAD

**Issue by quinthar** [Expensify/App#33574]
**Problem:**
Real world testing in NewDot #social has revealed that showing joined/left messages all the time is very noisy.  Further discussion uncovered that the value behind showing who has come/gone is limited to "private" chats, where the content of the chat is heavily dependent upon who is participating.  If people can come and go without awareness, it's hard to know what you are able to safely talk about.

A second, smaller problem is that the `joined the room` message says "room" even if what you joined isn't a room (eg, a thread, group, report, etc)

**Solution:**
- [x] Update the system message to a more general `David Barrett joined the chat` 
- [x] Add a corresponding `David Barrett left the chat` system message. 
- [x] Get rid of all "join" messages. See [this #vip-vsb slack thread](https://expensify.slack.com/archives/C066HJM2CAZ/p1705088355431289?thread_ts=1703797005.247439&cid=C066HJM2CAZ) for context.
- [x] Suppress `David Barrett left the chat` and `David Barrett invited Ben to the chat` messages when the person joining/leaving can "freely come and go" under their own control.  For example, the message should be suppressed if the person leaving/joining is ANY of the following:
  - [x] a member of the workspace, and this is a `Workspace` visibility chat
  - [x] an admin of the workspace, and this is an expense report on that workspace
  - [x] already a member of the parent chat (ie, once you invite an outsider to a chat, we don't re-announce them joining threads off of that chat)
  - [x] COULD FREELY BECOME a member of the parent chat even if they aren't currently (ie, a workspace member can freely join a thread even if they aren't in the parent `Workspace` visibility room, because they _could_ join it at any time)
- [x] When someone is explicitly removed, show the `David Barrett removed Ben from the chat` in *all* room types, as we want to signal to everyone that they were administratively removed.

**Comment by blimpich** [comment]
This is a priority for me, but I'm currently catching up on engineering chores after being on vacation. Will get to this as soon as I have a free cycle from chores.

**Comment by blimpich** [comment]
Updated the issue description to use checkboxes so that I can more clearly mark off the requirements with smaller PRs.

**Comment by blimpich** [comment]
Not overdue Melvin, actively working on this, finishing up chores this morning before getting to this in the afternoon.

**Comment by blimpich** [comment]
Not closing, there's a lot more to go on this PR.

**Comment by blimpich** [comment]
Update: making progress on leave message (2nd checkbox), logic is a bit hard to follow in web-expensify for how to do it. Will require changes in both auth and web.

**Comment by blimpich** [comment]
Worked on this today. I've read through a lot of the message PRs, the design doc, probed the code, and I still seem to be having difficulty getting "John Smith left the chat" to show up. I'd like to talk to @jasperhuangg but they seem to be OOO. Planning on messaging them when they get back. Currently confused as to how to get "left the chat" to work. Oddly tricky.

**Comment by jasperhuangg** [comment]
@blimpich do you have some draft PRs? Can take a look when I have some time. You're probably just missing a piece somewhere.

The code for sending report actions is pretty complicated in Web, I feel ya.

**Comment by blimpich** [comment]
@jasperhuangg just put them up!
web here: https://github.com/Expensify/Web-Expensify/pull/40507
auth here: https://github.com/Expensify/Auth/pull/9658

**Comment by jasperhuangg** [comment]
@blimpich took a look at your PRs, seems you have the right idea for the backend portion of this so far.

Did you check if the LeaveRoom actions are being received in Onyx on the client? Do you know how to do so?

I don't have my laptop at the moment but that would be the next thing I would check. If you don't know how to check local storage in the client #eng-chat can help you out with that.

If they're getting received in Onyx (they should be), there's probably some logic in the front end that's preventing them from displaying. It seems you didn't link an App PR, but one should be necessary to get these new actions to display. Check the places we render JoinRoom actions in the front end.

**Comment by jasperhuangg** [comment]
@blimpich If you add `LEAVE_ROOM: 'LEAVEROOM'` [here](https://github.com/Expensify/App/blob/753ce92ea9a1858dd2d576eeff692df527ea3ecb/src/CONST.ts#L599) it works. The logic [here](https://github.com/Expensify/App/blob/main/src/libs/ReportActionsUtils.ts#L365-L368) was preventing them from displaying.

<img width="667" alt="Screenshot 2024-01-13 at 2 01 53 PM" src="https://github.com/Expensify/App/assets/31285285/a202d017-6362-4a1f-bb41-47a189b3aba2">

(there are dupes because of the comment I left on your PR [here](https://github.com/Expensify/Auth/pull/9658#discussion_r1451235841))

**Comment by jasperhuangg** [comment]
<img width="1036" alt="Screenshot 2024-01-13 at 2 04 51 PM" src="https://github.com/Expensify/App/assets/31285285/fb19d80c-0f8a-427e-9900-5b13c9c11377">

@blimpich In Chrome if you go to the Inspector > Application > IndexedDB > OnyxDB > keyvaluepairs > and then you search for the key `reportActions_<reportID>` you can see a list of all the report actions in local storage for that client.

Really useful for when you want to check that Pusher updates are actually being received by a client. That would have really helped you see that front-end filtering was the issue, and nothing with the back-end.

**Comment by jasperhuangg** [comment]
@blimpich any update here? Looks like you're really close to getting your `left` message PRs done.

**Comment by blimpich** [comment]
@jasperhuangg I was OOO monday+tuesday, I will try to get to this today.

**Comment by blimpich** [comment]
Ah, I did not realize that there was filtering going on in the frontend. I did check the onyx DB and saw that the updates were getting sent to the frontend, but it didn't occur to me that we would be filtering the report actions on the frontend for visibility. Thank you @jasperhuangg for pointing out the needed change in `Expensify/App`. It might be worth updating this section of [the design doc](https://docs.google.com/document/d/1V4DwJjzx_Uxa47w6sjAfYTUthH8whdmMIWqp2FrsH5U/edit#heading=h.xy0fnppzt4qe) to include the changes to `App/src/CONST.ts` for posterity's sake.

**Comment by blimpich** [comment]
Made good progress today thanks to @jasperhuangg 🎉. Got code working for leave messages, PRs are up but still need to work out some tests / polish, then send out for review. Will get to that tomorrow.

**Comment by blimpich** [comment]
PRs are up and ready for review for "left the chat" message.

**Comment by blimpich** [comment]
Ah, found a small bug while manually testing the "left the chat" message PRs. [Details here](https://github.com/Expensify/Auth/pull/9658#issuecomment-1899506292). Need to squash that first and then hopefully will be good to go. PRs are all still out for review so that we can move quickly once the bug is squashed.

**Comment by blimpich** [comment]
Update: squashed bug, responded to feedback, added tests, re-requested reviews. Steady progress. ETA for leave messages is first half of next week, depending on how future reviews and QA go.

**Comment by blimpich** [comment]
Not overdue, calm down Melvin 🤖. I continue to work on this.

**Comment by blimpich** [comment]
Responded to feedback, back out for review for "left the chat" messages.

**Comment by blimpich** [comment]
Update: responded to minor bits of feedback, merged main into feature branches in order to fix failing tests unrelated to changes, waiting for approval/merge for leave messages.

**Comment by blimpich** [comment]
Update: waiting on PR approvals, starting to poke around at the next couple bullet points, which should go quicker now that I have a better understanding of how system messages work in the app.

**Comment by quinthar** [comment]
Great!  What's the ETA for the rest?

**Comment by blimpich** [comment]
Not entirely certain, today I got pulled away by chore work, but off to top of my head is two weeks, so feb 16th. Will change that if I end up moving quicker/slower.

**Comment by blimpich** [comment]
Updated the issue description slightly, added a bullet point to remove join messages based off this slack convo https://expensify.slack.com/archives/C066HJM2CAZ/p1705088355431289?thread_ts=1703797005.247439&cid=C066HJM2CAZ

**Comment by blimpich** [comment]
@jasperhuangg I'm in the process of removing some join message code in auth, since we don't really need the "x joined the chat" messages. See [this slack thread](https://expensify.slack.com/archives/C066HJM2CAZ/p1705088355431289?thread_ts=1703797005.247439&cid=C066HJM2CAZ) for context or if you feel we should keep them. But my question is, if we get rid of join messages, do we still need the join reportAction?

**Comment by blimpich** [comment]
Update: leave messages are very close to being in production. Only need a merge on this PR: https://github.com/Expensify/App/pull/34694

**Comment by blimpich** [comment]
Not overdue, working on this. Join messages have been removed, still waiting on PR approval / merge for last leave message PR https://github.com/Expensify/App/pull/34694

**Comment by blimpich** [comment]
Update: final leave message PR is in, box checked off ✅

PR for next bullet point is almost up. Need to finish writing some tests.

**Comment by quinthar** [comment]
Woohoo!  What's the ETA for completing the whole issue?

**Comment by blimpich** [comment]
Optimistically the 16th. We did have [a bug](https://github.com/Expensify/App/issues/35477) come up with leave messages, and so unknown bugs might delay if more crop up, but I think I can get the PR for the last bullet point done sometime next week, and hopefully merged and tidied up by the end of the week after that, so the 16th.

**Comment by blimpich** [comment]
2nd to last bullet point turned out to be easier to implement than expected and review went smoothly, so that is done. On to the last bullet point.

**Comment by blimpich** [comment]
Update: last bullet point was also much easier than expected. Final PR for this is here: https://github.com/Expensify/Web-Expensify/pull/40775. Once that is in this should be done.

The one sorta linked issue to this is https://github.com/Expensify/App/issues/35278. But that already has its own issue for it. I'm gonna try to see if there are any bigger wave issues to work on before picking that one up.

**Comment by blimpich** [comment]
Not overdue, was the weekend. Close to being done.

**Comment by blimpich** [comment]
https://github.com/Expensify/Web-Expensify/pull/40775 just merged and deployed to staging. Waiting for that to go to production and then for melvin to close this issue. Should be done.

**Comment by quinthar** [comment]
Woohoo!

**Comment by blimpich** [comment]
Closing this since last PR was deployed to production a week ago.
